### PR TITLE
Supported yolov4-tiny

### PR DESCRIPTION
**Pull request by l-bat**
Merge with extra: opencv/opencv_extra#770

relates #17666

<cut/>

```
force_builders=Linux32,Win32,Custom,Custom Win,Custom Mac
build_image:Custom=ubuntu-openvino-2020.2.0:16.04
build_image:Custom Win=openvino-2020.2.0
build_image:Custom Mac=openvino-2020.2.0

test_modules:Custom=dnn,python2,python3,java
test_modules:Custom Win=dnn,python2,python3,java
test_modules:Custom Mac=dnn,python2,python3,java

buildworker:Custom=linux-1
test_opencl:Custom=ON
test_bigdata:Custom=1
test_filter:Custom=*YOLO*:*VINO*:*ENet*:*route*

test_opencl:Custom Win=ON
test_bigdata:Custom Win=1
test_filter:Custom Win=*YOLO*:*VINO*:*ENet*:*route*

```

### CODE REVIEW

File: modules/dnn/src/darknet/darknet_io.cpp [hunk 1]
```diff
@@ -363,6 +363,38 @@ namespace cv {
                     fused_layer_names.push_back(last_layer);
                 }
 
+                void setSlice(int number_of_inputs, int *input_indexes, int split_size, int split_id)
```
Comment (dkurt) [inline modules/dnn/src/darknet/darknet_io.cpp:366]: Why not to accept `std::vector<int>& `?

File: modules/dnn/src/darknet/darknet_io.cpp [hunk 2]
```diff
@@ -725,10 +758,27 @@ namespace cv {
...
+                        if (num_splits > 1)
+                        {
+                            int group_id = getParam<int>(layer_params, "group_id");
+                            tensor_shape[0] /= num_splits;
+                            int split_size = tensor_shape[0] / layers_vec.size();
+                            setParams.setSlice(layers_vec.size(), layers_vec.data(), split_size, group_id);
+                            if (layers_vec.size() > 1)
+                            {
+                                for (size_t k = 0; k < layers_vec.size(); ++k)
+                                    layers_vec[k] += layers_vec.size();
```
Comment (dkurt) [inline modules/dnn/src/darknet/darknet_io.cpp:770]: For which layers we should increment indices?

Comment (l-bat) [reply modules/dnn/src/darknet/darknet_io.cpp:770]: For Concat after Slice

Comment (dkurt) [reply modules/dnn/src/darknet/darknet_io.cpp:770]: But why we have a single Slice layer but have to increase by number of it's outputs?

Comment (l-bat) [reply modules/dnn/src/darknet/darknet_io.cpp:770]: We increment layer ids, not number of outputs

File: modules/dnn/src/darknet/darknet_io.cpp [hunk 3]
```diff
@@ -363,6 +363,38 @@ namespace cv {
...
+                        lp.bottom_indexes.push_back(fused_layer_names.at(input_indexes[i]));
+                        net->layers.push_back(lp);
+                        fused_layer_names.push_back(lp.layer_name);
+                    }
+
+                    if (input_indexes.size() == 1)
+                    {
+                        last_layer = cv::format("slice_%d_%d", layer_id, input_indexes[0]);
+                        layer_id++;
+                    }
```
Comment (dkurt) [inline modules/dnn/src/darknet/darknet_io.cpp:395]: Can you please write in comments how this layer configuration works depends on number of inputs?

File: modules/dnn/src/darknet/darknet_io.cpp (hunk 2 above)
Comment (dkurt) [reply modules/dnn/src/darknet/darknet_io.cpp:770]: Yes, but it;s not clear for me that for a single layer (Slice) we need to add N indices (by number of Concat inputs?). I believe some comment in code about layer behavior will help to explain this.

File: modules/dnn/src/darknet/darknet_io.cpp (hunk 3 above)
Comment (l-bat) [reply modules/dnn/src/darknet/darknet_io.cpp:396]: Comments have been added

File: modules/dnn/src/darknet/darknet_io.cpp [hunk 4]
```diff
@@ -363,6 +363,39 @@ namespace cv {
...
+                        lp.layer_name = cv::format("slice_%d_%d", layer_id, input_indexes[i]);
+                        lp.layer_type = slice_param.type;
+                        lp.layerParams = slice_param;
+                        lp.bottom_indexes.push_back(fused_layer_names.at(input_indexes[i]));
+                        net->layers.push_back(lp);
+                        fused_layer_names.push_back(lp.layer_name);
+                    }
+
+                    // if input_indexes.size() > 1: Slice -> Concat, so we increment layer_id in Concat layer
+                    if (input_indexes.size() == 1)
```
Comment (dkurt) [inline modules/dnn/src/darknet/darknet_io.cpp:392]: Comment says: `if input_indexes.size() > 1  we increment layer_id`
Next line:
```cpp
if (input_indexes.size() == 1)
layer_id++;
```

File: modules/dnn/src/darknet/darknet_io.cpp [hunk 5]
```diff
@@ -363,6 +363,39 @@ namespace cv {
                     fused_layer_names.push_back(last_layer);
                 }
 
+                void setSlice(const std::vector<int>& input_indexes, int split_size, int split_id)
+                {
+                    cv::dnn::LayerParams slice_param;
+                    slice_param.name = "Slice-name";
```
Comment (dkurt) [inline modules/dnn/src/darknet/darknet_io.cpp:369]: is this required?

File: modules/dnn/src/darknet/darknet_io.cpp [hunk 6]
```diff
@@ -717,6 +750,7 @@ namespace cv {
                     {
                         std::string bottom_layers = getParam<std::string>(layer_params, "layers", "");
                         CV_Assert(!bottom_layers.empty());
+                        int num_splits = getParam<int>(layer_params, "groups", 1);
```
Comment (dkurt) [inline modules/dnn/src/darknet/darknet_io.cpp:753]: please rename to `groups`. Number of splits means that we split something

**Comment by rafcy** [comment]
Guys i have a question, will yolov4/tiny be supported in 4.X version as well? 
i am asking because using 3.4.x version the backend/traget_cuda are not available to use in 3.4.x version.

**Comment by alalek** [comment]
@rafcy See https://github.com/opencv/opencv/wiki/Branches on general branches policy.
I believe @YashasSamaga or other contributors will help to port changes for CUDA backend (we have ~one week window before 4.4.0 release).

File: modules/dnn/src/darknet/darknet_io.cpp [hunk 7]
```diff
@@ -725,10 +750,31 @@ namespace cv {
...
 
-                        if (layers_vec.size() == 1)
-                            setParams.setIdentity(layers_vec.at(0));
+                        if (groups > 1)
+                        {
+                            int group_id = getParam<int>(layer_params, "group_id", 0);
+                            tensor_shape[0] /= groups;
+                            int split_size = tensor_shape[0] / layers_vec.size();
+                            for (size_t k = 0; k < layers_vec.size(); ++k)
+                                setParams.setSlice(layers_vec[k], split_size, group_id, true);
```
Comment (dkurt) [inline modules/dnn/src/darknet/darknet_io.cpp:759]: Why `true` while it's not Darknet's layer? See `else if (layer_type == "region")` with Permute as reference

**Pull request merged**